lib/mainloop.c

```diff
@@ -398,6 +399,59 @@ main_loop_verify_config(GString *result, MainLoop *self)
   g_free(file_contents);
 }
 
+/* main_loop_config_monitor_init
+ * initializes configuration monitoring */
+
+/* callback */
+static void
+_on_file_modified(const DirectoryMonitorEvent *event, gpointer user_data)
+{
+  static gboolean ignore_first_time_created_event = TRUE;
+  MainLoop *self = (MainLoop *) user_data;
+
+  if (strcmp(event->full_path, resolvedConfigurablePaths.cfgfilename))
+    return;
+
+  if (ignore_first_time_created_event && (event->event_type == FILE_CREATED))
+    {
+      ignore_first_time_created_event = FALSE;
+      return;
+    }
+
+  if ((event->event_type == FILE_MODIFIED) || (event->event_type == FILE_CREATED))
```

Comment:
If user_data would store the last modification time of the cfgfile besides mainloop, then you could compare the modification time from `g_stat` here. That would eliminate the need of `mtime` parameter in `collection_comparator`.

Reply:
Yes sorry, I should have explained my problem first, before offering an alternative solution.

From clean code perspective I have difficulties in understanding the `CollectionComparator` mechanics the way it is implemented now. The original design and implementation was clean, it served the purpose of checking tracked (general purpose) values, and executed callbacks based on a value has newly added or removed from the tracked list. And I can also imagine a improved version with modification time: we have 3 callbacks, new, removed, modified, that can be called respectively, which is your current intention to implement. Storing `mtime` for this sake in `CollectionComparatorEntry` is also reasonable.

My problem is the way `mtime` is injected into `CollectionComparator` when calling `collection_comparator_add_value`. The current implementation passes `time_t modified`, which looks out of place.

Passing modified time during `..._add` adds too much freedom, callers allow to update modification time at their will, they can set to zero, they can set into the the past or the future, without control. To avoid this, users are highly suggested (forced) to create a function that calculates the modification time before calling `..._add`, resulting in a repetitive pattern.

In my opinion it would be much cleaner if `CollectionComparator` could be taught how it could detect the modification change (which could use even a method that is not based on mtime, but I would not go that far in this scope, as it would probably YAGNI).

And there are alternatives to achieve this.
- One would be to specify another callback for collection comparator (`gboolean entry_modified_p(...)`), that is used to check if `handle_modified_entry` should be called or not.
- Or we could say that `handle_modified_entry` is called unconditionally (when specified), and users responsibility to provide a callback that check if the modification and necessary, and early return otherwise. In that case however, updating mtime in the entry would be the user's responsibility. That means `CollectionComparatorEntry` should be passed instead of `const gchar *`. With that it would be elegant to store `gpointer modified_data` instead of `mtime modified`, but I can live without that. Still it would be nice to unify the api, that `handle_new_entry` etc signatures should be updated. With that said, the other option would be better.

For our purpose however, as only a single file modification is enough, that's why I proposed my alternative above, that the directory monitor callback could deal with the update, which would eliminate many complexities from above. And I am not sure mtime tracking would be used by other features in the future apart from auto reload.

About goal shifting, I would like to leave that decision to you. My request is to address specifying mtime in `..._add` problem somehow.